File: python_version/bot.py
"""
Bot pour la plateforme awale-arena
Format compatible:
- Entrée: mouvements au format "NX" ou "NTX" (N=numéro trou, X=R/B/T, T=transparent)
- Sortie: coup au même format
- Utilise MinMax AI de ai_algorithms.py
"""

import sys
import random
import time
from typing import List, Optional, Tuple
from game_rules import GameState, Color
from game_engine import GameEngine, MoveGenerator
from ai_algorithms import MinMaxBot


def main():
    """
    Fonction principale pour interfacer avec awale-arena
    Lire depuis stdin, traiter les mouvements, envoyer les réponses
    """
    # Récupérer le numéro du joueur (1 ou 2)
    my_player = int(sys.argv[1])  # 1 ou 2

    # Initialiser le jeu et le bot avec MinMax AI
    state = GameState()
    engine = GameEngine(state)
    bot = MinMaxBot(depth=1)  # Start with depth 1, will increase with iterative deepening

    # Boucle principale
    for line in sys.stdin:
        move_str = line.strip()

        # Traiter les signaux spéciaux
        if move_str == "END":
            break

        # Pour Player 2, attendre le premier coup de Player 1
        if move_str == "START" and my_player == 2:
            continue  # Attendre le prochain message

        if move_str != "START":
            # Appliquer le coup de l'adversaire
            hole, color, trans_as = parse_move(move_str)
            if hole and color:
                engine.play_move(hole, color, trans_as)

        # S'assurer que le joueur courant est correct
        state.current_player = my_player
        
        # Use the new find_best_move with internal timeout checking
        # This uses iterative deepening and only saves results from completed depths
        start_time = time.time()
        
        best_move = bot.find_best_move(state, my_player, timeout_ms=2950)
        
        # Print timing info to stderr (won't affect game protocol)
        total_time = time.time() - start_time
        sys.stderr.write(f"[Python] Move time: {total_time:.3f}s\n")
        sys.stderr.flush()

        if best_move:
            hole, color, trans_as = best_move
            my_move = format_move(hole, color, trans_as)
            
            # Appliquer notre coup
            engine.play_move(hole, color, trans_as)
        else:
            # Pas de coup disponible (ne devrait pas arriver)
            my_move = "NOMOVE"

        # Envoyer le coup à la plateforme
        sys.stdout.write(my_move + "\n")
        sys.stdout.flush()


def iterative_deepening_search(bot, state, player, time_limit=2.0):
    """
    Iterative deepening with strict time limit (must return before 3s).
    
    Algorithm:
    1. Calculate depth 1, save the result
    2. Calculate depth 2, if valid, save it (replaces previous)
    3. Continue increasing depth while time allows
    4. Always return the best result found
    """
    start_time = time.time()
    best_move = None
    final_depth = 0
    
    # Start at depth 1 and keep going deeper (max depth 10)
    for depth in range(1, 11):
        depth_start = time.time()
        
        # Calculate move at current depth
        bot.depth = depth
        move = bot.get_move(state, player)
        
        depth_time = time.time() - depth_start
        
        # Save the result if valid
        if move is not None:
            best_move = move
            final_depth = depth
        
        # Check remaining time
        elapsed = time.time() - start_time
        remaining = time_limit - elapsed
        
        # Estimate next depth time (use 8x as branching can be high)
        # Stop if next depth would likely exceed remaining time
        estimated_next = depth_time * 8
        if estimated_next > remaining:
            break
    
    # Print timing info to stderr (won't affect game protocol)
    total_time = time.time() - start_time
    sys.stderr.write(f"[Python] Move time: {total_time:.3f}s | Depth: {final_depth}\n")
    sys.stderr.flush()
    
    return best_move


def parse_move(move_str: str) -> Tuple[Optional[int], Optional[Color], Optional[Color]]:
    """
    Parse un mouvement au format "NX" ou "NTX"
    Retourne (hole, color, transparent_as)
    """
    try:
        move_str = move_str.strip().upper()
        
        if 'T' in move_str[1:-1]:  # Transparent au milieu (ex: "3TR", "3TB")
            hole = int(move_str[:-2])
            trans_as_str = move_str[-1]
            color = Color.TRANSPARENT
            trans_as = Color.RED if trans_as_str == 'R' else Color.BLUE
            return (hole, color, trans_as)
        else:  # Normal (ex: "3R", "4B")
            hole = int(move_str[:-1])
            color_str = move_str[-1]
            color = Color.RED if color_str == 'R' else Color.BLUE
            return (hole, color, None)
    except:
        return (None, None, None)


def format_move(hole: int, color: Color, trans_as: Optional[Color]) -> str:
    """
    Formate un mouvement pour l'envoi
    """
    if color == Color.TRANSPARENT:
        return f"{hole}T{trans_as.value}"
    else:
        return f"{hole}{color.value}"


if __name__ == "__main__":
    main()

File: c_version/bot.cpp
#include "game_rules.h"
#include "game_engine.h"
#include "ai_algorithms.h"
#include <iostream>
#include <string>
#include <sstream>
#include <tuple>
#include <optional>
#include <chrono>

struct ParsedMove
{
    int hole;
    Color color;
    std::optional<Color> trans_as;
    bool valid;
};

ParsedMove parse_move(const std::string &move_str)
{
    ParsedMove result = {0, Color::RED, std::nullopt, false};

    try
    {
        std::string upper_move = move_str;
        for (auto &c : upper_move)
            c = std::toupper(c);

        if (upper_move.length() >= 3 && upper_move[upper_move.length() - 2] == 'T')
        {
            result.hole = std::stoi(upper_move.substr(0, upper_move.length() - 2));
            result.color = Color::TRANSPARENT;

            char trans_as_char = upper_move[upper_move.length() - 1];
            result.trans_as = (trans_as_char == 'R') ? Color::RED : Color::BLUE;
            result.valid = true;
        }
        else if (upper_move.length() >= 2)
        {
            result.hole = std::stoi(upper_move.substr(0, upper_move.length() - 1));

            char color_char = upper_move[upper_move.length() - 1];
            result.color = (color_char == 'R') ? Color::RED : Color::BLUE;
            result.valid = true;
        }
    }
    catch (...)
    {
        result.valid = false;
    }

    return result;
}

std::string format_move(int hole, Color color, std::optional<Color> trans_as)
{
    std::stringstream ss;
    ss << hole;

    if (color == Color::TRANSPARENT && trans_as.has_value())
    {
        ss << "T" << (trans_as.value() == Color::RED ? "R" : "B");
    }
    else
    {
        ss << (color == Color::RED ? "R" : "B");
    }

    return ss.str();
}

int main(int argc, char *argv[])
{
    if (argc < 2)
    {
        return 1;
    }

    int my_player = (std::string(argv[1]) == "JoueurA") ? 1 : 2;

    GameState state;
    GameEngine engine(&state);
    MinMaxBot bot(1);
    int move_count = 0;
    std::string last_move;

    std::string line;

    auto check_game_over = [&]() -> std::string
    {
        int score_j1 = state.captured_seeds[1];
        int score_j2 = state.captured_seeds[2];

        if (score_j1 >= 49 || score_j2 >= 49)
        {
            return "RESULT " + last_move + " " + std::to_string(score_j1) + " " + std::to_string(score_j2);
        }

        if (move_count >= 400)
        {
            return "RESULT LIMIT " + std::to_string(score_j1) + " " + std::to_string(score_j2);
        }

        int total_on_board = 0;
        for (int i = 1; i <= 16; i++)
        {
            total_on_board += state.holes[i][Color::RED] + state.holes[i][Color::BLUE] + state.holes[i][Color::TRANSPARENT];
        }
        if (total_on_board < 10)
        {
            return "RESULT " + last_move + " " + std::to_string(score_j1) + " " + std::to_string(score_j2);
        }

        return "";
    };

    while (std::getline(std::cin, line))
    {
        line.erase(0, line.find_first_not_of(" \t\r\n"));
        line.erase(line.find_last_not_of(" \t\r\n") + 1);

        if (line == "START" && my_player == 2)
        {
            continue;
        }

        if (line != "START")
        {
            ParsedMove parsed = parse_move(line);
            if (parsed.valid)
            {
                if (parsed.color == Color::TRANSPARENT && parsed.trans_as.has_value())
                {
                    engine.playMove(parsed.hole, parsed.color, parsed.trans_as.value(), true);
                }
                else
                {
                    engine.playMove(parsed.hole, parsed.color, Color::RED, false);
                }
                move_count++;
                last_move = line;

                std::string result = check_game_over();
                if (!result.empty())
                {
                    std::cout << result << std::endl;
                    break;
                }
            }
        }

        state.current_player = my_player;

        auto start_time = std::chrono::steady_clock::now();

        Move best_move = bot.findBestMove(state, my_player, std::chrono::milliseconds(2000));

        auto end_time = std::chrono::steady_clock::now();
        long total_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count();

        std::string my_move;

        if (best_move.valid)
        {
            std::optional<Color> trans_opt = best_move.use_transparent ? std::optional<Color>(best_move.transparent_as) : std::nullopt;
            my_move = format_move(best_move.hole, best_move.color, trans_opt);

            engine.playMove(best_move.hole, best_move.color,
                            best_move.transparent_as, best_move.use_transparent);
            move_count++;
            last_move = my_move;

            std::cout << my_move << std::endl;
            std::cout.flush();

            std::string result = check_game_over();
            if (!result.empty())
            {
                std::cout << result << std::endl;
                break;
            }
        }
        else
        {
            int score_j1 = state.captured_seeds[1];
            int score_j2 = state.captured_seeds[2];
            std::cout << "RESULT " << last_move << " " << score_j1 << " " << score_j2 << std::endl;
            break;
        }
    }

    return 0;
}
